### DISCUSSION TITLE
New developer onboarding

### DISCUSSION THREAD

**Discussion by tekenny** [w7sst/MorseRunner#107] [General]
Disclaimer: It's been a few decades since I've learned (and taught) Pascal. So I'm a bit rusty but I've programmed in many different languages since then so I'm sure I'll get up to speed quickly.

I'm thinking of contributing a new file named README-newdev.md (or maybe README-dev-onboard.md) where I'll document how a new contributor can onboard aka get started. As an alternative maybe this belongs in CONTRIBUTING.md but I thought a separate README for contributing developers would be useful. #FeedbackWelcomed

The following are the initial getting started questions I have. Excuse me if this info already exist elsewhere but I haven't seen it in this repo.

1 - Which Pascal? Should I be using "Free Pascal" or some other version of Pascal?
2 - Is there a recommended/suggested IDE? Up to personal preference? Use the Free Pascal IDE?
3 - Are there any existing unit tests? I was hoping to find a tests subdirectory but that doesn't exist.
4 - How do I run the source code the first time to test changes? I ran Main.pas within Free Pascal IDE and got a fatal error "Can't find unit Graphics used by Main"
5 - How is the software built for consumption by the user community? 
6 - Are there automated nightly builds of the code? If not is that on the roadmap?

I'm sure I'll have more questions... Once I get fully capable I'll mine this thread an my learnings for developer onboard documentation.

Thanks for assisting with my dumb newb questions!
73, Tom NJ2DX

**Comment by w7sst** [comment]
Hi Tom,
As you know, there are no dumb questions. I first learned Pascal in the late 70's while in college and haven't touched it since until this spring when I started looking at this project. I was impressed with how much OO support is in the language and supporting libraries.

Let me try to answer some of your questions.

1. Which Pascal - Yes, I think this is all based on FPC.
2. IDEs - The original projects were using Lazarus.
- I have both Lazarus and Delphi installed on my laptop. This github repository has branches from both IDEs.
- This project is using the free Delphi - Community Edition. I currently am using version 10.4. This Delphi CE is free to non-profit projects and small teams - I think we qualify.
- One of the challenges we have is how to support the two IDEs. One question we will want to tackle as a team is choosing one. There is currently active development using both IDEs. I don't know if it's possible to provide concurrent support for both (meaning same code runs on both). I have picked Delphi for now and will look in Lazarus again in the future. I'm working with @n2ic on N1MM integration and he is currently using Lazarus for his builds. I have not started a conversation yet with Steve (N2IC)  on where we go long term on this issue.
3. Units tests - I wish. I'd like to see some added, but none at this point. Current code is still highly coupled, so it will take some refactoring to separate the lower-level modules. I'm open to any ideas you or others might have.
4. Running source code - When running from the IDE, I just put the IDE In debug mode (set to compile the debug version) and then hit F9 to start application under the debugger. I can then set breakpoints and debug as needed.
5. Production builds - I currently build a Release version of the product, copy the necessary files to a separate directory, and then zip the directory and upload it to the [Releases](https://github.com/w7sst/MorseRunner/releases) section of our github page. I have not built any makefiles to help with this. I notice that one fork by zmetzing included a Makefile for compiling outside of the IDE. We might be able to learn something from his project. Reach out to him if you have any questions. It appears he was working on a linux port, so I'm interested in pulling his stuff in sometime in the future.
6. Nightly builds - good idea. I have read about this, but don't know how to access a Delphi or Lazarus compile environment. We need someone with experiencing in setting something like this up. So yes, please add this idea to the roadmap for next spring and add a `Help Wanted` Issue asking for help setting up a nightly build environment - that would be cool.

The convention we are following is to put the support pages under our `.github/` directory. This is where the CONTRIBUTING.md file resides. I would suggest adding a new `.github/DEVELOPERS.md` file and then link to it from `README.md`.

Thank you for starting to look at the developer side of things. I had a vision in the spring to see if a community-based approach to MR support would work. David, @ct7aup, and Scott, @scotthibbs, joined me in the summer and we have grow the repository to where it is now, but there is still more to be done. It's good to see interest in this project slowly starting considering we have not made any major announcements for it yet. We wanted to get the software stable before we do so - which includes finishing some of this refactoring I'm working on right now. I'm guess the announcement might slip until January since I'd like to slow down on this over the month of December.

73, Mike W7SST

**Comment by w7sst** [comment]
Hi @BH1SCW, @f6fvy, @ct7aup, @jm1dxh, @jr8ppg, @n2ic, @scotthibbs,

Tom, @tekenny, recently joined our group. He is starting to collect and document new developer information and compile procedures.

Do have any notes and comments about your developing, compilation, or install procedures that could help Tom document this for the MR project.

One area that is tricky with this project, and has always confused me, is handling of the VCL (visual component library) into either IDE. Any notes or insight you have will help.

Thank you and 73,
Mike W7SST

**Comment by n2ic** [comment]
Sorry for my lack of activity. The last thing I have done is get the N1MM+
version of MR working with 64 bit Lazarus.

I am no lover of Lazarus ! I only used it because there was nothing else
under $500 when I started working on MR. If the Delphi Community Edition
works well, I would be happy to switch. It would be even more wonderful if
there was Pascal/Delphi support in Visual Studio, but that's asking too
much.

73,
Steve, N2IC

**Comment by tekenny** [comment]
I'd like to use Visual Studio as well but I don't know if or how hard it would be to get running in it...
